Support operator strategies in MT

vidaldid-rte opened the pull request:
**Please check if the PR fulfills these requirements**
<!-- please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes -->
- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**Does this PR already have an issue describing the problem?**
No

**What kind of change does this PR introduce?**
Today, mutli-thread security analysis does not support operator strategy,
because each thread verifies that a contincency exists for every action. But each thread has only a sublist of contingencies so
this fatally leads to errors.

With this PR, the operator strategy checks are performed before running the treads. So that:
   - exceptions thrown for missing action ids are sent to the launching thread
   - matching contingencies for operator strategy contingency contexts are performed before the contingency partitioning


**Does this PR introduce a breaking change or deprecate an API?**
- [ ] Yes
- [X] No

Hadrien-Godard commented on src/main/java/com/powsybl/openloadflow/sa/AbstractSecurityAnalysis.java:
```diff
@@ -148,19 +148,34 @@ SecurityAnalysisReport runSync(SecurityAnalysisParameters securityAnalysisParame
+                    .filter(o -> !hasValidContingency(o, contingencyIds))
+                    .findAny()
+                    .ifPresent(o -> throwOperatorStrategyWithoutAction(o));
+            // Check action ids to report exce ption to the main thread
```
Typo here: "exception" instead of "exce ption"

vidaldid-rte replied:
Done

Hadrien-Godard commented on src/main/java/com/powsybl/openloadflow/sa/AbstractSecurityAnalysis.java:
```diff
@@ -474,30 +493,57 @@ protected static Map<String, Action> indexActionsById(List<Action> actions) {
-                // check actions IDs exists
+            if (hasValidContingency(operatorStrategy, contingencyIds)) {
+                findMissingActionId(operatorStrategy, actionIds)
+                        .ifPresent(id -> throwMissingdOperatorStrategyAction(operatorStrategy, id));
```
throwMissingdOperatorStrategyAction is used even if throwIfInvalidOperatorStrategy values "false". Is it ok?

vidaldid-rte replied:
Good point. It does not hurt but the verification is a waste of time since it was done before. Il will correct.

vidaldid-rte replied:
Done

vidaldid-rte commented on src/main/java/com/powsybl/openloadflow/sa/AbstractSecurityAnalysis.java:
```diff
@@ -474,30 +493,57 @@ protected static Map<String, Action> indexActionsById(List<Action> actions) {
+        return Optional.empty();
+    }
+
+    private static void throwOperatorStrategyWithoutAction(OperatorStrategy operatorStrategy) {
```
Done

Hadrien-Godard reviewed (approved):
I approve the changes.
Just one thought, maybe we should solve the Sonar issue and rearrange the new boolean parameter as a security analysis parameter?

vidaldid-rte commented:
> I approve the changes. Just one thought, maybe we should solve the Sonar issue and rearrange the new boolean parameter as a security analysis parameter?

Good point @Hadrien-Godard . I removed the parameter from the runSimulations methods and replaced it by a logic on thread count. Does it work for you ?

Hadrien-Godard commented:
> Member

Yes perfect! Thanks!

Hadrien-Godard commented:
> > I approve the changes. Just one thought, maybe we should solve the Sonar issue and rearrange the new boolean parameter as a security analysis parameter?
> 
> Good point @Hadrien-Godard . I removed the parameter from the runSimulations methods and replaced it by a logic on thread count. Does it work for you ?

Yes perfect! Thanks!
